### PR TITLE
Implement Inventory.getPlugin(), pretty much everywhere.

### PR DESCRIPTION
**Pull request by JBYoshi**
**SpongeCommon** | [SpongeForge](https://github.com/SpongePowered/SpongeForge/pull/1521)

Fixes SpongePowered/SpongeCommon#1222.

### CODE REVIEW

File: src/main/java/org/spongepowered/common/mixin/core/item/inventory/TraitInventoryAdapter.java [hunk 1]
```diff
@@ -132,4 +136,33 @@ protected final EmptyInventory emptyInventory() {
...
+                    .orElseThrow(() -> new AssertionError("Missing plugin " + pluginId + " for block " + id));
+        }
+        if (base instanceof Entity) {
+            String id = ((Entity) this).getType().getId();
+            String pluginId = id.substring(0, id.indexOf(":"));
+            return Sponge.getPluginManager().getPlugin(pluginId)
+                    .orElseThrow(() -> new AssertionError("Missing plugin " + pluginId + " for entity " + id + " (" + getClass().getName() + ")"));
+        }
+        if (base instanceof SpongeUser) {
+            return SpongeImpl.getMinecraftPlugin();
```
Comment (Zidane) [inline src/main/java/org/spongepowered/common/mixin/core/item/inventory/TraitInventoryAdapter.java:162]: I wonder if this should be the SpongeAPI plugin instead..

File: src/main/java/org/spongepowered/common/item/inventory/EmptyInventoryImpl.java [hunk 2]
```diff
@@ -261,7 +261,7 @@ public void notify(Object source, InventoryEventArgs eventArgs) {
 
     @Override
     public PluginContainer getPlugin() {
-        return null; // TODO?
+        return parent.getPlugin();
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/item/inventory/EmptyInventoryImpl.java]: `this.`

**Review by kashike — CHANGES_REQUESTED** [review]

File: src/main/java/org/spongepowered/common/mixin/core/item/inventory/TraitInventoryAdapter.java (hunk 1 above)
Comment (JBYoshi) [reply src/main/java/org/spongepowered/common/mixin/core/item/inventory/TraitInventoryAdapter.java:162]: I decided to use the Minecraft plugin for consistency with the `Player` inventory; they're conceptually the same, so IMO they should have the same owner. I'm open to changing it, however.

**Comment by JBYoshi** [comment]
Can this be merged?

**Comment by Faithcaio** [comment]
Can you add testplugin?

File: src/main/java/org/spongepowered/common/mixin/core/item/inventory/TraitInventoryAdapter.java [hunk 3]
```diff
@@ -128,4 +136,33 @@ protected final EmptyInventory emptyInventory() {
...
+        }
+
+        // TODO
+        SpongeImpl.getLogger().warn("Unknown plugin for " + this);
```
Comment (Zidane) [inline src/main/java/org/spongepowered/common/mixin/core/item/inventory/TraitInventoryAdapter.java]: Can you finish this TODO?

File: src/main/java/org/spongepowered/common/mixin/core/item/inventory/TraitInventoryAdapter.java [hunk 4]
```diff
@@ -128,4 +136,33 @@ protected final EmptyInventory emptyInventory() {
...
+    public PluginContainer getPlugin() {
+        Object base = this;
+        if (this instanceof CarriedInventory) {
+            Optional<?> carrier = ((CarriedInventory<?>) this).getCarrier();
+            if (carrier.isPresent()) {
+                base = carrier.get();
+            }
+        }
+        if (base instanceof TileEntity) {
+            String id = ((TileEntity) this).getBlock().getType().getId();
```
Comment (Zidane) [inline src/main/java/org/spongepowered/common/mixin/core/item/inventory/TraitInventoryAdapter.java]: I concur with @Faithcaio about the testing plugin for this part alone.

**Comment by JBYoshi** [comment]
@Zidane I finished up the TODO.

@Faithcaio I got the test plugin ready. Just open and close any inventory and you should see the credits pop up. (And by credits I mean the name of the plugin that created the inventory. :wink:)

**Comment by parlough** [comment]
You are missing the license on the test plugin, but otherwise looks good.

**Comment by JBYoshi** [comment]
*facepalm*
Time to rework my Git hooks...

**Comment by Zidane** [comment]
@JBYoshi 

No worries, I can take care of the license when I pull.

**Comment by Zidane** [comment]
Merged in: https://github.com/SpongePowered/SpongeCommon/commit/4f6471aae9b031835bb16706a38ca9dcd30e6091

**Pull request closed without merging**